les/clientpool.go

```diff
@@ -0,0 +1,591 @@
+		}
+	}
+	// client accepted, finish setting it up
+	if nb != nil {
+		delete(f.negBalanceMap, address)
+		f.negBalanceQueue.Remove(nb.queueIndex)
+	}
+	if e.priority {
+		e.balanceTracker.addCallback(balanceCallbackZero, 0, func() { f.balanceExhausted(id) })
+	}
+	f.connectedMap[id] = e
+	f.connectedQueue.Push(e)
+	f.connectedCapacity += e.capacity
+	if e.capacity != f.freeClientCap {
+		e.peer.updateCapacity(e.capacity)
+	}
+	log.Debug("Client accepted", "address", address)
+	return e
+}
+
+// unregisterPeer implements peerSetNotify
+func (f *clientPool) unregisterPeer(p *peer) {
+	f.disconnect(p)
+}
+
+// disconnect should be called when a connection is terminated. If the disconnection
+// was initiated by the pool itself using disconnectFn then calling disconnect is
+// not necessary but permitted.
+func (f *clientPool) disconnect(p clientPeer) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	if f.closed {
+		return
+	}
+	address := p.freeClientId()
+	id := p.ID()
+	// Short circuit if the peer hasn't been registered.
+	e := f.connectedMap[id]
+	if e == nil {
+		log.Debug("Client not connected", "address", address, "id", peerIdToString(id))
+		return
+	}
+	f.dropClient(e, f.clock.Now(), false)
+}
+
+// dropClient removes a client from the connected queue and finalizes its balance.
+// If kick is true then it also initiates the disconnection.
+func (f *clientPool) dropClient(e *clientInfo, now mclock.AbsTime, kick bool) {
+	if _, ok := f.connectedMap[e.id]; !ok {
+		return
+	}
+	f.finalizeBalance(e, now)
+	f.connectedQueue.Remove(e.queueIndex)
+	delete(f.connectedMap, e.id)
+	f.connectedCapacity -= e.capacity
+	if kick {
+		log.Debug("Client kicked out", "address", e.address)
+		f.removePeer(e.id)
+	} else {
+		log.Debug("Client disconnected", "address", e.address)
+	}
+}
+
+// finalizeBalance stops the balance tracker, retrieves the final balances and
+// stores them in posBalanceQueue and negBalanceQueue
+func (f *clientPool) finalizeBalance(c *clientInfo, now mclock.AbsTime) {
+	c.balanceTracker.stop(now)
+	pos, neg := c.balanceTracker.getBalance(now)
+	pb := f.getPosBalance(c.id)
+	pb.value = pos
+	if neg < 1 {
+		neg = 1
+	}
+	nb := &negBalance{address: c.address, queueIndex: -1, logValue: int64(math.Log(float64(neg))*fixedPointMultiplier) + f.logOffset(now)}
+	f.negBalanceMap[c.address] = nb
+	f.negBalanceQueue.Push(nb, -nb.logValue)
+	if f.negBalanceQueue.Size() > f.queueLimit {
+		nn := f.negBalanceQueue.PopItem().(*negBalance)
+		delete(f.negBalanceMap, nn.address)
+	}
+}
+
+// balanceExhausted callback is called by balanceTracker when positive balance is exhausted.
+// It revokes priority status and also reduces the client capacity if necessary.
+func (f *clientPool) balanceExhausted(id enode.ID) {
+	f.lock.Lock()
+	defer f.lock.Unlock()
+
+	c := f.connectedMap[id]
+	if c == nil || !c.priority {
+		return
+	}
+	c.priority = false
+	if c.capacity != f.freeClientCap {
+		f.connectedCapacity += f.freeClientCap - c.capacity
```

Comment:
Does it work? 

I have a test snippet

```
a := uint64(64)
a += uint64(10) - uint64(32)
fmt.Println(a)
```

And the output is `constant -22 overflows uint64`

Reply:
Sure, on the machine code level there is absolutely no difference between signed and unsigned addition/subtraction. If you specify constants then the compliler will check for overflows but this is not the case here. I could typecast it to int64 but typecasting is also an implicit assumption about value range and I do not think it would make the code look better. If it an established standard though (or at least done everywhere else in our codebase) then I can add a typecast in these cases.  In that case I have to check the rest of my code too though because I think I do this regularly :)

Reply:
One other example is prque.Prque which uses a "wrap-around" priority scheme where the comparison is int64(a-b) in order to allow perpetually moving offsets. Binary integers are modulo values technically and I have always considered it perfectly "legal" to think about them and use them that way.